common/core/hwclayer.cpp

```diff
@@ -229,6 +229,41 @@ void HwcLayer::SufaceDamageTransfrom() {
 
   if (!surface_damage_.empty() &&
       ((source_crop_.left == 0) && (source_crop_.top == 0))) {
+#ifdef RECT_DAMAGE_TRACING
+    IRECTDAMAGETRACE("Calculating Damage for layer[%d]", z_order_);
+    IRECTDAMAGETRACE("Surface_damage (LTWH): %d, %d, %d, %d",
+                     surface_damage_.left, surface_damage_.top,
+                     (surface_damage_.right - surface_damage_.left),
+                     (surface_damage_.bottom - surface_damage_.top));
+    IRECTDAMAGETRACE(
+        "Original current_rendering_damage_ (LTWH): %d, %d, %d, %d",
+        current_rendering_damage_.left, current_rendering_damage_.top,
+        (current_rendering_damage_.right - current_rendering_damage_.left),
+        (current_rendering_damage_.bottom - current_rendering_damage_.top));
+    IRECTDAMAGETRACE("display_frame_ (LTWH): %d, %d, %d, %d",
+                     display_frame_.left, display_frame_.top,
+                     (display_frame_.right - display_frame_.left),
+                     (display_frame_.bottom - display_frame_.top));
+    IRECTDAMAGETRACE("source_crop_ (LTWH): %f, %f, %f, %f", source_crop_.left,
+                     source_crop_.top, (source_crop_.right - source_crop_.left),
+                     (source_crop_.bottom - source_crop_.top));
+#endif
+    int display_width = display_frame_.right - display_frame_.left;
+    int display_height = display_frame_.bottom - display_frame_.top;
+    int source_width = source_crop_.right - source_crop_.left;
+    int source_height = source_crop_.bottom - source_crop_.top;
+    int surface_width = surface_damage_.right - surface_damage_.left;
+    int surface_height = surface_damage_.bottom - surface_damage_.top;
+
+    if (display_width != surface_width && display_height != surface_height) {
```

Comment:
This judge is not about the actually display resolution, right?
According to your description,  We need to get the real resolution?

Reply:
In the bug case, the surface Rect is calculated with 1080P coordination when calculating the rect size for 4k/2k display. the size of surface Rect should be enlarged from the source size (1080P) to the display size (2k/4k).

Reply:
@Shao-Feng  How about if display is smaller than 1080P?

Reply:
That is why we use “!=” rather than ">" here. when the surface size is not equal with the display, it will be enlarged or reduced.

Reply:
This patch is in-consistent with commit description.
Acording to this patch, the damage rect is not zoomed to display frame in SF. Instead HWC shuld zoom it, it is not about the display resolution, 2K or 4K right
@yhe39  Please also comment
Let's test the cases we know to make sure no reg

Reply:
In playing 480P or 4K video with mosaic mode, the display size is already 3840*1080, and the surface is also calculated based on  3840*1080. for 480p, the size of video surface is almost 1920*1080. with this judge, the size will not be changed.

Reply:
Chunge is running a reg scope testing for all 7 bugs which patch is related with this method. we can wait for her result

Reply:
LGTM, i think this patch could be merge if no regression is found. thanks!